pop/service/service.go

```diff
@@ -224,47 +466,451 @@ func (s *Service) CheckConfig(req *network.Envelope) {
+			hashLocal, err := final.Hash()
+			if err != nil {
+				log.Error("VerifyMerge: hash computation failed")
+				return false
+			}
+			hashReceived, err := finalReceived.Hash()
+			if err != nil {
+				log.Error("VerifyMerge: hash computation failed")
+				return false
+			}
+			if !bytes.Equal(hashLocal, hashReceived) {
+				log.Error("VerifyMerge: hashes Received and Local are not equal", s.ServerIdentity())
+				return false
+			}
+
+			// check that merge config is completed in merge
+			if len(stmtsMap) != len(final.Desc.Parties) {
+				log.Error("VerifyMerge: length of Merge and Merge Config are not equal", s.ServerIdentity())
+				return false
+			}
+			for _, mergeStmt := range stmtsMap {
+				status := final.VerifyMergeStatement(mergeStmt)
+				if status < PopStatusOK {
+					log.Error("VerifyMerge: Received non valid FinalStatement", s.ServerIdentity())
+					return false
+				}
 			}
+			finals = append(finals, final)
 		}
-		s.data.Final.Attendees[i] = nil
 	}
-	s.data.Final.Attendees = na
+
+	if !found {
+		log.Error("VerifyMerge: no party from merge was found locally")
+		return false
+	}
+
+	m := &merge{stmtsMap, true}
+	var syncData *sync
+	if syncData, ok = s.syncs[string(final.Desc.Hash())]; !ok {
+		log.Error("VerifyMerge: No sync data with given hash")
+		return false
+	}
+
+	// Merge fields
+	locs := make([]string, 0)
+	Roster := &onet.Roster{}
+	na := make([]abstract.Point, 0)
+	for _, f := range m.statementsMap {
+		// although there must not be any intersection
+		// in attendies list it's better to check it
+		// not simply extend the list
+		na = unionAttendies(na, f.Attendees)
+		Roster = unionRoster(Roster, f.Desc.Roster)
+		locs = append(locs, f.Desc.Location)
+	}
+	sortAll(locs, Roster.List, na)
+	final.Desc.Location = strings.Join(locs, DELIMETER)
+	final.Merged = true
+	final.Desc.Roster = Roster
+	final.Attendees = na
+
+	// check that Msg is valid
+	hashLocal, err := final.Hash()
+	if err != nil {
+		log.Error("VerifyMerge: hash computation failed")
+		return false
+	}
+
+	if !bytes.Equal(hashLocal, Msg) {
+		log.Error("Msg is invalid", s.ServerIdentity())
+		return false
+	}
+
+	// update local data
+	newHash := string(final.Desc.Hash())
+	s.data.Finals[newHash] = final
+	s.data.merges[newHash] = m
+	s.syncs[newHash] = syncData
+	m.statementsMap = make(map[string]*FinalStatement)
+	m.statementsMap[newHash] = final
+
+	// change final statement for all parties which were going to merge
+	// to backward compatibility with orgs, who can't get hash of new final statement
+	for _, f := range finals {
+		// but signature on this conodes will be invalid
+		// because it's impossible to save signature on old hashes
+		s.data.Finals[string(f.Desc.Hash())] = final
+		// there is no need to support consistency of syncData and merge
+		// for old parties because their finalStatements are rewritten
+	}
+
+	s.save()
+	return true
+}
+
+/* --------------Propagation function-------------- */
```

Comment:
you can remove all those `----`-comments, we don't use it in our other packages.